frontend/src/components/main-page/groups-tree-page/common/dependencies/visualization/sankey-links.ts

```diff
@@ -0,0 +1,190 @@
+): RawLink[] {
+  const result: RawLink[] = [];
+
+  let servicesAndGroups: Array<RegularGroupNode | ServiceNode>;
+  if (buildConfig.mode === 'build-around-group') {
+    servicesAndGroups = getDiagonalRelativesOfGroup(
+      buildConfig.group,
+      rootGroup,
+    );
+  } else {
+    servicesAndGroups = extractAllServices(rootGroup);
+  }
+
+  for (const singleServiceOrGroup of servicesAndGroups) {
+    const APIsAndEvents = getAllAPIsAndEvents(singleServiceOrGroup);
+
+    /*
+      The "natural" flow of Events if from producer to consumer:
+      (SomeProducer) --> (Event) --> (SomeConsumer)
+    */
+    for (const producedEvent of APIsAndEvents.producedEvents) {
+      result.push({
+        type: 'from-service-or-group',
+        from: singleServiceOrGroup,
+        to: producedEvent,
+      });
+    }
+
+    for (const consumedEvent of APIsAndEvents.consumedEvents) {
+      result.push({
+        type: 'to-service-or-group',
+        from: consumedEvent,
+        to: singleServiceOrGroup,
+      });
+    }
+
+    /*
+      The "natural" flow of APIs is from consumer to provider:
+      (SomeConsumer) --> (API) --> (SomeProducer)
+    */
+    for (const consumedAPI of APIsAndEvents.consumedAPIs) {
+      result.push({
+        type: 'from-service-or-group',
+        from: singleServiceOrGroup,
+        to: consumedAPI,
+      });
+    }
+
+    for (const providedAPI of APIsAndEvents.providedAPIs) {
+      result.push({
+        type: 'to-service-or-group',
+        from: providedAPI,
+        to: singleServiceOrGroup,
+      });
+    }
+  }
+
+  return result;
+}
+
+/**
+ * The "diagonal relatives" (or "diagonal neighbors") of a node are:
+ * 1. The direct descendants of this node.
+ * 2. The siblings of this node.
+ * 3. Siblings of all ancestors of this node.
+ *
+ * Assume the following tree (for simplicity without distinction between groups and services):
+ *
+ * ```
+ *              Root
+ *         /     |       \
+ *    A          B           C
+ *  / | \      / | \       / | \
+ * D  E  F    G  H  I     J  K  L
+ * |  |  |    |  |  |     |  |  |
+ * M  N  O    P  Q  R     S  T  U
+ * ```
+ *
+ * The diagonal relatives of node A are D, E, F, B, and C.
+ * Explanation:
+ * - Rule 1 gives us D, E, and F.
+ * - Rule 2 gives us B and C.
+ * - Rule 3 does not give us any nodes.
+ *
+ * The diagonal relatives of node D are M, E, F, B, and C.
+ * Explanation:
+ * - Rule 1 gives us M.
+ * - Rule 2 gives us E and F.
+ * - Rule 3 gives us B and C.
+ *
+ *
+ * Remarks:
+ * - One could see this as having selected nodes A, B, and C, and then having "punched into A so that a dent has been formed". *
+ * - We only take inner nodes into account, i.e. the computation is undefined for leaf nodes.
+ * - In our Service Docs model, we have groups and services. Here, we say that the siblings of a group node X are all sibling groups of X, as well as all services directly owned by X.
+ */
+function getDiagonalRelativesOfGroup(
```

Comment:
There is probably a proper graph-theoretical term for this. But I did not find a suitable one, so I decided to call if "Diagonal Relatives", because it feels a bit like having drawn a diagonal line across the tree.